Relax the rule that a color only is allowed in the second position

JockeJarre opened the pull request:
Relax the rule that a color has to be only in the second position in DOF configuration file.

As described in the ([DOF documentation](https://directoutput.github.io/DirectOutput/inifiles.html#inifiles_generalpara) the **Color name** or **Hex color definition** has to be placed in as the second value (e.g. S48 Blue). (Trigger Color)

I have been working with mberneis on the DOF Config tool to allow Templates to be defined. It means that even if there isn't many MX definitions available, you can build MX definitions based on the RGB or non color counterparts. (and maybe other things as well) 
For more details see [VP Universe](https://vpuniverse.com/forums/topic/11769-dof-config-tool-template-support/)

We do not want this template feature to know _to much_ about the DOF internals, rather simple text handling, which means if you add a color to a DOF configuration which has more than a trigger, your color addition would be placed further to the right, which make DOF fail.

With this change DOF would allow the following:

```
On Blink fu500 fd600 Red AT0 AL0
```

The part starting with **Red...** is added on top of the Start button configuration using a template, hence creating a MX configuration using the same trigger as the non colored one.

Even something like this would be "ok"

```
ON Blue AT0 AL0 AW100 AH100 Red
```

It means the **Red** *override* the **Blue** configuration. -> changing a preferred color...

Of course the old style configuration files will still work with this change.

JockeJarre committed:
Relax the rule that a color configuration has to be only in the second position in DOF configuration file.

As described in the [DOF documentation](https://directoutput.github.io/DirectOutput/inifiles.html#inifiles_generalpara) the **Color name** or
**Hex color definition** has to be placed in as the second value (e.g. S48 Blue). (Trigger Color)

I have been working with mberneis on the DOF Config tool to allow Templates to be defined.
It means that even if there isn't many MX definitions available, you can build MX definitions based on the RGB or non color counterparts. (and maybe other things as well) For more details see [VP Universe](https://vpuniverse.com/forums/topic/11769-dof-config-tool-template-support/)

We do not want this template feature to know to much about the DOF internals, rather simple text handling, which means if you add a color to a DOF configuration which has more than a trigger, your color addition would be placed further to the right, which make DOF fail.

With this change DOF would allow the following:

```
On Blink fu500 fd600 Red AT0 AL0
```

The part starting with **Red...** is added on top of the Start button configuration using a template. Creating a MX configuration using the same trigger.

Even something like this would be "ok"

```
ON Blue AT0 AL0 AW100 AH100 Red
```

It means the **Red** *override* the **Blue** configuration. -> changing a preferred color...

Of course the old style configuration files will still work with this change.

mjrgh commented on DirectOutput/LedControl/Loader/TableConfigSetting.cs:
```diff
@@ -745,10 +745,16 @@ public void ParseSettingData(string SettingData)
                 }
-                else if (PartNr == 0)
+                // if Parts[PartNr] starts with capital letter and the rest small caps letters or underscore
+                else if (Parts[PartNr].Length > 2 && char.IsUpper(Parts[PartNr][0]) && Parts[PartNr].Skip(1).All(c => (char.IsLower(c) || c == '_')))
```
This really should be a regex.  I think something like this would be a lot clearer and more maintainable: `if (Regex.IsMatch(Parts[PartNr], @"^[A-Z_][a-z_]#")) `

But before doing that, are you really sure that "a mixed-case word starting with a capital letter" is actually a good way to say "this is a color"?  That seems like an awfully ad hoc and fragile rule to me.  It's certainly not a convention that anyone else uses, so no one's just going to come up with it on their own intuitively.  I know DOF configuration is already a horrible mess, but do we really want to make it even worse by adding another weird opaque rule like this?

Could we instead just look in the ColorConfigList to see if the token matches something in the list?  If the answer is Yes, it's a color; if the answer is No, and it's not a '#' HTML-style color code, we end up in the last case, where we log an error that says (essentially) "unrecognized token".  Maybe that error message could be improved to suggest that the token needs to be either one of the keywords or the name of a color.  I suppose you could still use the mixed-case heuristic to decide what error message to show - if it's mixed-case, the error message could say something like "Unrecognized token {0}; it doesn't match any known keyword or color name (was this meant to be a color name?)".  But that's pretty much the same error to show whether it's mixed case or not, so I'm talking myself out of the heuristic being useful even for that.

mjrgh commented on DirectOutput/LedControl/Loader/TableConfigSetting.cs:
```diff
@@ -745,10 +745,16 @@ public void ParseSettingData(string SettingData)
+                    ColorName = Parts[PartNr].ToUpper();
+                }
+                else if (Parts[PartNr].Length >= 7 && Parts[PartNr].Length <= 9 && Parts[PartNr][0] == '#' && Parts[PartNr].Substring(1).All(c => Uri.IsHexDigit(c)) )
+                {
```
This would also be improved by changing to a regex: `if (Regex.IsMatch(Parts[PartNr], @""^#[0-9A-Fa-f]{6,8)$))`.  (I'm not actually sure that we should enforce the format so strictly, though.  It might be better diagnostic-wise to make this simply `if (Regex.IsMatch(Parts[PartNr], "@^#"))`, since it doesn't seem that a string starting with '#' could be anything other than an html-style color code; and then, only after matching that, add a second test against the full regex, and log an error if it doesn't match ("Invalid '#' color code format, #rrggbb or #rrggbbaa required").

mjrgh reviewed (commented):
See my code review comments - let me know what you think.

JockeJarre committed:
Improve error messages for #HTML-style hex color values and stricter color check using regular expression
Check possible color names longer than 2 chars and sanity check using regular expression.
Improve error message on non identified color

JockeJarre commented:
I think it is now improved from the initial state, which meant that everything not handled in the "normal" parameters would be used as a color (when in the first position). Now there is at least a simple sanity check using the regular expression.
Any string which is handed over as a color which isn't found will be properly logged.

JockeJarre committed:
Also make Trigger uppercase

mjrgh commented on DirectOutput/LedControl/Loader/TableConfigSetting.cs:
```diff
@@ -745,10 +746,21 @@ public void ParseSettingData(string SettingData)
+                }
+                // if Parts[PartNr] contains only capital and small caps letters or underscore we assume a color
+                else if (Parts[PartNr].Length > 2 && Regex.IsMatch(Parts[PartNr], @"^[A-Za-z_]+$"))
+                {
```
Is it possible here to check that the string is one of the known color names, before assuming it's a color based on the lexical pattern alone?  I definitely prefer this to the original change where we had the first-letter-capitalized rule to identify a color, but this kind of goes to the other extreme, where everything that's not recognized as one of the other keywords is assumed to be a color.  That might be really confusing when the user intended to enter a keyword, but made a slight typo.  I think the ideal here would be to recognize the defined color names specifically, alongside the other keywords, so that we can log an error if the input is neither a color nor a keyword.  An error message is always better than an incorrect assumption, since it can be hard to figure out why the program is making the weird assumption when you're looking at something you THINK is a properly spelled keyword.

It seems to me that we do have a color table that we can look at, but maybe the initialization order is such that the color table isn't populated at the point where we parse this part of the file.  But if we can successfully look at the color table at that stage, I think it would be really helpful to users to make this test conditional on the token matching an entry in the color table, and then falling through to the "Unrecognized keyword" case if not.

JockeJarre replied:
Hi Michael, yes of course you are right. It would be better. The reason I had go to the other extreme, is that (as you stated) the quality of the color entries was far to bad. I got a lot of errors in the log.
Using this now a little too loose, we still filter many of the errors out, and those looking like a color (not having any numeric values embedded) will now be logged in the Configurator.cs (where the colors are loaded). I changed the log message to actually log the text coming in.
Through these new messages I found 2 handful of wrong entries in the DB.

mjrgh replied:
> The reason I had go to the other extreme, is that (as you stated) the quality of the color entries was far to bad. I got a lot of errors in the log.

I take it you mean that you got too many **spurious** errors (errors that aren't really errors)?  Or do you mean that the DOF database is so riddled with **actual** errors that it's just too much noise to log them all?  If it's legitimate errors, I wonder if it would be worth coordinating with the Config Tool people to clean up the database so that we could tighten up the diagnostics here.  It seems like both projects would benefit from that.

mjrgh reviewed (commented):
This looks good.  One thing before I merge it - see comment at TableConfigSettings.cs 762.

JockeJarre commented:
With my change we can now see errors of the Kind "ON RedA100" (The color stuck to some other parameter missing a space) The RedA100 went to the Color check, which did't log the errornous text coming in. The current version beleive everthing on second spot which doesn't fit any parameter is a color.
Apart from that we have many colors written in different caps (yellow, YELLOW and correct Yellow), so my first intepretation with having one Capital sign + all small caps together with "_" didn't hold. 
With the current regex the kind of error like RedA100, will directly be logged and all those which aren't captured here, will be logged at the point where it tries to interpret it as a color.
I am also discussing with the dev of the new DOF config tool, the best would be to add a sanity check already in there, to minimize the bad quality already at the source.

mjrgh commented:
> With the current regex the kind of error like RedA100, will directly be logged and all those which aren't captured here, will be logged at the point where it tries to interpret it as a color.

Ah, that sounds good then.  I was thinking that no errors would be logged at all.  No need to log the error twice.  Is there enough information still available where the error does get logged (as you say, the point where it interprets it as a color) to pinpoint the location of the error?  If not, maybe it would be better to check the color and log the error in the parser, and then suppress the error in the other place, since it's already being reported.  (Or just don't set the color entry that triggers the second error in the first place.)

> I am also discussing with the dev of the new DOF config tool, the best would be to add a sanity check already in there, to minimize the bad quality already at the source.

Glad to hear it!  That would be great, since users who just generate their configs from the DOF database shouldn't have to figure out whey their log files are full of error messages that have nothing to do with anything they did.

JockeJarre commented:
This is an example of the new log since we are a little more strict of what is a color (isn't passed as a color at all):
`2025.06.21 13:35:33.575	EXCEPTION: Message: Exception --> Cant parse the part BlueAT85 of the ledcontrol table config setting W16 BlueAT85  AH15 F150 ADR AS250 BLINK 800 BPW13 W400`

This is an example of the updated log of a string passed as a color which could not be reqognised as one: (just a fake test by me)
```
2025.06.21 13:35:34.685	Warning: No valid color definition "BREEN" found for area effect. Skipped setting 1 in column 4 for LedWizEqivalent number 30.
2025.06.21 13:35:34.689	Table config loading finished: romname=Testtable, tablename=Testtable (Original 2025)
```
Earlier it didn't log the **"BREEN"** at all.

So with both these new updates we have improved logging and we allow having color information not only in position no.1

JockeJarre commented:
While we are working on the directoutputconfig parser. 

If you have a directoutputconfig30.ini file: ( I have intensionally reduced the lenght of the lines)

```
[Config DOF]
# "Rom","Magnasave Left MX","","","Flipper Button MX","","" ...
a7x,On Blue,On Gold AT0 AL0 AW100 AH100 ...
aar,0,0,0,ON Blue AT0 AL0 AW100 AH100 ...
```
We have two configuration lines showing only two "MX" led configurations each. What is anoying me is that when a led config is empty, it has to use 3 "columns" of the "CSV" like data, but when the led config is active it uses only one. 
I have tried to find where this happens in the code, but wasn't successful yet. 
Do you happen to know where this happens?
I have also tried replacing three 0 into one Off, but then the whole parser breaks.

mjrgh commented:
> Do you happen to know where this happens?

Sorry, afraid not.

JockeJarre commented:
> This is an example of the new log since we are a little more strict of what is a color (isn't passed as a color at all): `2025.06.21 13:35:33.575 EXCEPTION: Message: Exception --> Cant parse the part BlueAT85 of the ledcontrol table config setting W16 BlueAT85 AH15 F150 ADR AS250 BLINK 800 BPW13 W400`
> 
> This is an example of the updated log of a string passed as a color which could not be reqognised as one: (just a fake test by me)
> 
> ```
> 2025.06.21 13:35:34.685	Warning: No valid color definition "BREEN" found for area effect. Skipped setting 1 in column 4 for LedWizEqivalent number 30.
> 2025.06.21 13:35:34.689	Table config loading finished: romname=Testtable, tablename=Testtable (Original 2025)
> ```
> 
> Earlier it didn't log the **"BREEN"** at all.
> 
> So with both these new updates we have improved logging and we allow having color information not only in position no.1

Sorry, distracted with my second question above...
What do you think? Should we merge the changes?

mjrgh commented:
> What do you think? Should we merge the changes?

Yes, I'm planning to - I'll do it when I have a chance, hopefully in the next few days.
